File: compatibility/pg/java/PGTest.java
import java.io.*;
import java.sql.*;
import java.util.List;
import java.util.LinkedList;

public class PGTest {
    public static class Tests {
        private Connection conn;
        private Statement st;
        private List<Test> tests = new LinkedList<>();

        public void connect(String ip, int port, String user, String password) {
            try {
                String url = "jdbc:postgresql://" + ip + ":" + port + "/postgres";
                conn = DriverManager.getConnection(url, user, password);
                st = conn.createStatement(ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY);
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        }

        public void disconnect() {
            try {
                st.close();
                conn.close();
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        }

        public void addTest(String query, String[][] expectedResults) {
            tests.add(new Test(query, expectedResults));
        }

        public boolean runTests() {
            for (Test test : tests) {
                if (!test.run()) {
                    return false;
                }
            }
            return true;
        }

        public void readTestsFromFile(String filename) {
            try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
                String line;
                while ((line = br.readLine()) != null) {
                    if (line.trim().isEmpty()) continue;
                    String query = line;
                    List<String[]> results = new LinkedList<>();
                    while ((line = br.readLine()) != null && !line.trim().isEmpty()) {
                        results.add(line.split(","));
                    }
                    String[][] expectedResults = results.toArray(new String[0][]);
                    addTest(query, expectedResults);
                }
            } catch (IOException e) {
                e.printStackTrace();
                System.exit(1);
            }
        }

        public class Test {
            private String query;
            private String[][] expectedResults;

            public Test(String query, String[][] expectedResults) {
                this.query = query;
                this.expectedResults = expectedResults;
            }

            public boolean run() {
                try {
                    System.out.println("Running test: " + query);
                    if (!st.execute(query)) {
                        if (expectedResults.length != 0) {
                            System.err.println("Expected " + expectedResults.length + " rows, got 0");
                            return false;
                        }
                        System.out.println("Returns 0 rows");
                        return true;
                    }
                    ResultSet rs = st.getResultSet();
                    if (rs.getMetaData().getColumnCount() != expectedResults[0].length) {
                        System.err.println("Expected " + expectedResults[0].length + " columns, got " + rs.getMetaData().getColumnCount());
                        return false;
                    }
                    int rows = 0;
                    while (rs.next()) {
                        int cols = 0;
                        for (String expected : expectedResults[rows]) {
                            String result = rs.getString(cols + 1).trim();
                            if (!expected.equals(result)) {
                                System.err.println("Expected:\n'" + expected + "'");
                                System.err.println("Result:\n'" + result + "'");
                                return false;
                            }
                            cols++;
                        }
                        rows++;
                    }

                    System.out.println("Returns " + rows + " rows");
                    if (rows != expectedResults.length) {
                        System.err.println("Expected " + expectedResults.length + " rows");
                        return false;
                    }
                    return true;
                } catch (SQLException e) {
                    System.err.println(e.getMessage());
                    return false;
                }
            }
        }
    }

    public static void main(String[] args) {
        if (args.length < 5) {
            System.err.println("Usage: java PGTest <ip> <port> <user> <password> <testFile>");
            System.exit(1);
        }

        Tests tests = new Tests();
        tests.connect(args[0], Integer.parseInt(args[1]), args[2], args[3]);
        tests.readTestsFromFile(args[4]);

        if (!tests.runTests()) {
            tests.disconnect();
            System.exit(1);
        }
        tests.disconnect();
    }
}

File: compatibility/pg/php/pg_test.php
<?php

class PGTest {
    public static function main($args) {
        if (count($args) != 5) {
            echo "Usage: php " . $args[0] . " <ip> <port> <user> <password> <testFile>\n";
            exit(1);
        }

        $tests = new Tests();
        $tests->connect($args[0], intval($args[1]), $args[2], $args[3]);
        $tests->readTestsFromFile($args[4]);

        if (!$tests->runTests()) {
            $tests->disconnect();
            exit(1);
        }
        $tests->disconnect();
    }
}

class Tests {
    private $conn;
    private $tests = [];

    public function connect($ip, $port, $user, $password) {
        try {
            $url = "pgsql:host=$ip;port=$port;dbname=postgres";
            $this->conn = new PDO($url, $user, $password);
        } catch (PDOException $e) {
            throw new RuntimeException($e->getMessage());
        }
    }

    public function disconnect() {
        $this->conn = null;
    }

    public function addTest($query, $expectedResults) {
        $this->tests[] = new Test($query, $expectedResults);
    }

    public function runTests() {
        foreach ($this->tests as $test) {
            if (!$test->run($this->conn)) {
                return false;
            }
        }
        return true;
    }

    public function readTestsFromFile($filename) {
        $file = fopen($filename, "r");
        if (!$file) {
            throw new RuntimeException("Failed to open test data file");
        }

        while (($line = fgets($file)) !== false) {
            $line = trim($line);
            if (empty($line)) continue; // Skip empty lines
            $query = $line;
            $expectedResults = [];
            while (($line = fgets($file)) !== false) {
                $line = trim($line);
                if (empty($line)) break; // End of expected results for this query
                $expectedResults[] = explode(",", $line);
            }
            $this->addTest($query, $expectedResults);
        }

        fclose($file);
    }
}

class Test {
    private $query;
    private $expectedResults;

    public function __construct($query, $expectedResults) {
        $this->query = $query;
        $this->expectedResults = $expectedResults;
    }

    public function run($conn) {
        try {
            echo "Running test: " . $this->query . "\n";
            $st = $conn->prepare($this->query);
            $st->execute();
            if ($st->columnCount() == 0) {
                if (count($this->expectedResults) != 0) {
                    echo "Expected " . count($this->expectedResults) . " rows, got 0\n";
                    return false;
                }
                echo "Returns 0 rows\n";
                return true;
            }
            $rows = $st->fetchAll(PDO::FETCH_NUM);
            if (count($rows[0]) != count($this->expectedResults[0])) {
                echo "Expected " . count($this->expectedResults[0]) . " columns, got " . count($rows[0]) . "\n";
                return false;
            }
            foreach ($rows as $i => $row) {
                $row = array_map('strval', $row);
                foreach ($row as $j => $value) {
                    $value = trim($value);
                    $expectedValue = trim($this->expectedResults[$i][$j]);
                    if ($value !== $expectedValue) {
                        echo "Expected: " . $expectedValue . ", got: " . $value . "\n";
                        return false;
                    }
                }
            }
            echo "Returns " . count($rows) . " rows\n";
            if (count($rows) != count($this->expectedResults)) {
                echo "Expected " . count($this->expectedResults) . " rows\n";
                return false;
            }
            return true;
        } catch (PDOException $e) {
            echo $e->getMessage() . "\n";
            return false;
        }
    }
}

PGTest::main(array_slice($argv, 1));

?>